File: mlx-sys/shim/src/quantization.cc
#include "cxx_mlx_shim/quantization.h"
#include "cxx_mlx_shim/shim_helpers.h"

#include <chrono>
#include <stdexcept>

#include "mlx/transforms.h"

namespace cxx_mlx {

using helpers::decode_stream_or_device;
using helpers::dtype_from_repr;
using helpers::opt_arr;
using helpers::opt_dtype;
using helpers::opt_i;

// ===== QuantizeResult =====

QuantizeResult::QuantizeResult(std::vector<mlx::core::array> data)
    : arrays_(std::move(data)), taken_(arrays_.size(), false) {}

std::unique_ptr<MlxArray> QuantizeResult::take_at(size_t idx) {
  if (idx >= arrays_.size()) {
    throw std::runtime_error("QuantizeResult::take_at: idx out of range");
  }
  if (taken_[idx]) {
    throw std::runtime_error("QuantizeResult::take_at: already taken at idx");
  }
  taken_[idx] = true;
  return std::make_unique<MlxArray>(std::move(arrays_[idx]));
}

size_t quantize_result_count(const QuantizeResult& r) { return r.count(); }

std::unique_ptr<MlxArray> quantize_result_take_at(QuantizeResult& r, size_t idx) {
  return r.take_at(idx);
}

// ===== quantize =====

std::unique_ptr<QuantizeResult> quantize(
    const MlxArray& w,
    bool has_group_size, int32_t group_size,
    bool has_bits, int32_t bits,
    rust::Str mode,
    const MlxArray* global_scale,
    bool has_target, bool is_device_only, uint8_t device_type, int32_t stream_index) {
  auto target = decode_stream_or_device(has_target, is_device_only, device_type, stream_index);
  auto result = mlx::core::quantize(
      w,
      opt_i(has_group_size, group_size),
      opt_i(has_bits, bits),
      std::string(mode),
      opt_arr(global_scale),
      target);
  return std::make_unique<QuantizeResult>(std::move(result));
}

// ===== dequantize =====

std::unique_ptr<MlxArray> dequantize(
    const MlxArray& w, const MlxArray& scales,
    const MlxArray* biases,
    bool has_group_size, int32_t group_size,
    bool has_bits, int32_t bits,
    rust::Str mode,
    const MlxArray* global_scale,
    bool has_dtype, uint8_t dtype_repr,
    bool has_target, bool is_device_only, uint8_t device_type, int32_t stream_index) {
  auto target = decode_stream_or_device(has_target, is_device_only, device_type, stream_index);
  return std::make_unique<MlxArray>(mlx::core::dequantize(
      w, scales, opt_arr(biases),
      opt_i(has_group_size, group_size),
      opt_i(has_bits, bits),
      std::string(mode),
      opt_arr(global_scale),
      opt_dtype(has_dtype, dtype_repr),
      target));
}

// ===== quantized_matmul =====

std::unique_ptr<MlxArray> quantized_matmul(
    const MlxArray& x, const MlxArray& w, const MlxArray& scales,
    const MlxArray* biases,
    bool transpose,
    bool has_group_size, int32_t group_size,
    bool has_bits, int32_t bits,
    rust::Str mode,
    bool has_target, bool is_device_only, uint8_t device_type, int32_t stream_index) {
  auto target = decode_stream_or_device(has_target, is_device_only, device_type, stream_index);
  return std::make_unique<MlxArray>(mlx::core::quantized_matmul(
      x, w, scales, opt_arr(biases),
      transpose,
      opt_i(has_group_size, group_size),
      opt_i(has_bits, bits),
      std::string(mode),
      target));
}

std::unique_ptr<MlxArray> quantized_matmul_batch_isolated(
    const MlxArray& x, const MlxArray& w, const MlxArray& scales,
    const MlxArray* biases,
    bool transpose,
    bool has_group_size, int32_t group_size,
    bool has_bits, int32_t bits,
    rust::Str mode,
    bool has_target, bool is_device_only, uint8_t device_type, int32_t stream_index) {
  auto target = decode_stream_or_device(has_target, is_device_only, device_type, stream_index);
  return std::make_unique<MlxArray>(mlx::core::quantized_matmul_batch_isolated(
      x, w, scales, opt_arr(biases),
      transpose,
      opt_i(has_group_size, group_size),
      opt_i(has_bits, bits),
      std::string(mode),
      target));
}

std::unique_ptr<MlxArray> quantized_matmul_product_stable(
    const MlxArray& x, const MlxArray& w, const MlxArray& scales,
    const MlxArray* biases,
    bool transpose,
    bool has_group_size, int32_t group_size,
    bool has_bits, int32_t bits,
    rust::Str mode,
    bool has_target, bool is_device_only, uint8_t device_type, int32_t stream_index) {
  auto target = decode_stream_or_device(has_target, is_device_only, device_type, stream_index);
  return std::make_unique<MlxArray>(mlx::core::quantized_matmul_product_stable(
      x, w, scales, opt_arr(biases),
      transpose,
      opt_i(has_group_size, group_size),
      opt_i(has_bits, bits),
      std::string(mode),
      target));
}

rust::Vec<double> quantized_matmul_bench_ms(
    const MlxArray& x, const MlxArray& w, const MlxArray& scales,
    const MlxArray* biases,
    bool transpose,
    bool has_group_size, int32_t group_size,
    bool has_bits, int32_t bits,
    rust::Str mode,
    size_t runs,
    bool has_target, bool is_device_only, uint8_t device_type, int32_t stream_index) {
  auto target = decode_stream_or_device(has_target, is_device_only, device_type, stream_index);
  auto maybe_biases = opt_arr(biases);
  auto maybe_group_size = opt_i(has_group_size, group_size);
  auto maybe_bits = opt_i(has_bits, bits);
  auto mode_string = std::string(mode);

  rust::Vec<double> timings;
  timings.reserve(runs);
  for (size_t i = 0; i < runs; ++i) {
    auto started = std::chrono::steady_clock::now();
    auto y = mlx::core::quantized_matmul(
        x, w, scales, maybe_biases,
        transpose,
        maybe_group_size,
        maybe_bits,
        mode_string,
        target);
    mlx::core::eval(std::vector<mlx::core::array>{y});
    mlx::core::synchronize();
    auto elapsed = std::chrono::steady_clock::now() - started;
    timings.push_back(std::chrono::duration<double, std::milli>(elapsed).count());
  }
  return timings;
}

// ===== qqmm =====

std::unique_ptr<MlxArray> qqmm(
    const MlxArray& x, const MlxArray& w,
    const MlxArray* w_scales,
    bool has_group_size, int32_t group_size,
    bool has_bits, int32_t bits,
    rust::Str mode,
    const MlxArray* global_scale_x,
    const MlxArray* global_scale_w,
    bool has_target, bool is_device_only, uint8_t device_type, int32_t stream_index) {
  auto target = decode_stream_or_device(has_target, is_device_only, device_type, stream_index);
  return std::make_unique<MlxArray>(mlx::core::qqmm(
      x, w, opt_arr(w_scales),
      opt_i(has_group_size, group_size),
      opt_i(has_bits, bits),
      std::string(mode),
      opt_arr(global_scale_x),
      opt_arr(global_scale_w),
      target));
}

// ===== gather_qmm =====

std::unique_ptr<MlxArray> gather_qmm(
    const MlxArray& x, const MlxArray& w, const MlxArray& scales,
    const MlxArray* biases,
    const MlxArray* lhs_indices,
    const MlxArray* rhs_indices,
    bool transpose,
    bool has_group_size, int32_t group_size,
    bool has_bits, int32_t bits,
    rust::Str mode,
    bool sorted_indices,
    bool has_target, bool is_device_only, uint8_t device_type, int32_t stream_index) {
  auto target = decode_stream_or_device(has_target, is_device_only, device_type, stream_index);
  return std::make_unique<MlxArray>(mlx::core::gather_qmm(
      x, w, scales, opt_arr(biases),
      opt_arr(lhs_indices), opt_arr(rhs_indices),
      transpose,
      opt_i(has_group_size, group_size),
      opt_i(has_bits, bits),
      std::string(mode),
      sorted_indices,
      target));
}

std::unique_ptr<MlxArray> from_fp8(
    const MlxArray& x, uint8_t dtype_repr,
    bool has_target, bool is_device_only, uint8_t device_type, int32_t stream_index) {
  auto target = decode_stream_or_device(has_target, is_device_only, device_type, stream_index);
  return std::make_unique<MlxArray>(mlx::core::from_fp8(x, dtype_from_repr(dtype_repr), target));
}

std::unique_ptr<MlxArray> to_fp8(
    const MlxArray& x,
    bool has_target, bool is_device_only, uint8_t device_type, int32_t stream_index) {
  auto target = decode_stream_or_device(has_target, is_device_only, device_type, stream_index);
  return std::make_unique<MlxArray>(mlx::core::to_fp8(x, target));
}

}  // namespace cxx_mlx

File: mlx-sys/src/bridge/quantization.rs
//! Bridge for MLX quantization subsystem.
//!
//! Quantize returns std::vector<array>, which cxx 1.0 doesn't support
//! as Vec<UniquePtr<T>>. Wrapped as opaque QuantizeResult with
//! count() + take_at(idx) free functions. Single-use semantics:
//! take_at(idx) twice throws (matches P2c take_by_name pattern).
//!
//! Optional encodings:
//! - Option<i32>   → (bool has_value, i32 value)  (P2b rope pattern)
//! - Option<Dtype> → (bool has_dtype, u8 dtype_repr)
//! - Option<&Array>→ *const MlxArray (nullptr = None)  (P2b/P2c pattern)
//! - &str mode     → rust::Str  (P2b sdpa pattern)
//!
//! Each fn carries 4 trailing `StreamOrDevice` args (P5.7) — same encoding
//! as the array bridge: `(has_target, is_device_only, device_type, stream_index)`.

#[allow(clippy::missing_safety_doc, clippy::too_many_arguments)]
#[cxx::bridge(namespace = "cxx_mlx")]
pub mod ffi {
    unsafe extern "C++" {
        include!("cxx_mlx_shim/quantization.h");

        type MlxArray = crate::bridge::array::ffi::MlxArray;
        type QuantizeResult;

        // ===== quantize result accessors =====
        fn quantize_result_count(r: &QuantizeResult) -> usize;
        fn quantize_result_take_at(
            r: Pin<&mut QuantizeResult>,
            idx: usize,
        ) -> Result<UniquePtr<MlxArray>>;

        // ===== quantize =====
        unsafe fn quantize(
            w: &MlxArray,
            has_group_size: bool,
            group_size: i32,
            has_bits: bool,
            bits: i32,
            mode: &str,
            global_scale: *const MlxArray,
            has_target: bool,
            is_device_only: bool,
            device_type: u8,
            stream_index: i32,
        ) -> Result<UniquePtr<QuantizeResult>>;

        // ===== dequantize =====
        unsafe fn dequantize(
            w: &MlxArray,
            scales: &MlxArray,
            biases: *const MlxArray,
            has_group_size: bool,
            group_size: i32,
            has_bits: bool,
            bits: i32,
            mode: &str,
            global_scale: *const MlxArray,
            has_dtype: bool,
            dtype_repr: u8,
            has_target: bool,
            is_device_only: bool,
            device_type: u8,
            stream_index: i32,
        ) -> Result<UniquePtr<MlxArray>>;

        // ===== quantized_matmul =====
        unsafe fn quantized_matmul(
            x: &MlxArray,
            w: &MlxArray,
            scales: &MlxArray,
            biases: *const MlxArray,
            transpose: bool,
            has_group_size: bool,
            group_size: i32,
            has_bits: bool,
            bits: i32,
            mode: &str,
            has_target: bool,
            is_device_only: bool,
            device_type: u8,
            stream_index: i32,
        ) -> Result<UniquePtr<MlxArray>>;

        unsafe fn quantized_matmul_batch_isolated(
            x: &MlxArray,
            w: &MlxArray,
            scales: &MlxArray,
            biases: *const MlxArray,
            transpose: bool,
            has_group_size: bool,
            group_size: i32,
            has_bits: bool,
            bits: i32,
            mode: &str,
            has_target: bool,
            is_device_only: bool,
            device_type: u8,
            stream_index: i32,
        ) -> Result<UniquePtr<MlxArray>>;

        unsafe fn quantized_matmul_product_stable(
            x: &MlxArray,
            w: &MlxArray,
            scales: &MlxArray,
            biases: *const MlxArray,
            transpose: bool,
            has_group_size: bool,
            group_size: i32,
            has_bits: bool,
            bits: i32,
            mode: &str,
            has_target: bool,
            is_device_only: bool,
            device_type: u8,
            stream_index: i32,
        ) -> Result<UniquePtr<MlxArray>>;

        unsafe fn quantized_matmul_product_stable_affine8_wide(
            x: &MlxArray,
            w: &MlxArray,
            scales: &MlxArray,
            biases: *const MlxArray,
            transpose: bool,
            has_group_size: bool,
            group_size: i32,
            has_bits: bool,
            bits: i32,
            mode: &str,
            has_target: bool,
            is_device_only: bool,
            device_type: u8,
            stream_index: i32,
        ) -> Result<UniquePtr<MlxArray>>;

        unsafe fn quantized_matmul_bench_ms(
            x: &MlxArray,
            w: &MlxArray,
            scales: &MlxArray,
            biases: *const MlxArray,
            transpose: bool,
            has_group_size: bool,
            group_size: i32,
            has_bits: bool,
            bits: i32,
            mode: &str,
            runs: usize,
            has_target: bool,
            is_device_only: bool,
            device_type: u8,
            stream_index: i32,
        ) -> Result<Vec<f64>>;

        // ===== qqmm =====
        unsafe fn qqmm(
            x: &MlxArray,
            w: &MlxArray,
            w_scales: *const MlxArray,
            has_group_size: bool,
            group_size: i32,
            has_bits: bool,
            bits: i32,
            mode: &str,
            global_scale_x: *const MlxArray,
            global_scale_w: *const MlxArray,
            has_target: bool,
            is_device_only: bool,
            device_type: u8,
            stream_index: i32,
        ) -> Result<UniquePtr<MlxArray>>;

        // ===== gather_qmm =====
        unsafe fn gather_qmm(
            x: &MlxArray,
            w: &MlxArray,
            scales: &MlxArray,
            biases: *const MlxArray,
            lhs_indices: *const MlxArray,
            rhs_indices: *const MlxArray,
            transpose: bool,
            has_group_size: bool,
            group_size: i32,
            has_bits: bool,
            bits: i32,
            mode: &str,
            sorted_indices: bool,
            has_target: bool,
            is_device_only: bool,
            device_type: u8,
            stream_index: i32,
        ) -> Result<UniquePtr<MlxArray>>;

        // ===== FP8 =====
        fn from_fp8(
            x: &MlxArray,
            dtype_repr: u8,
            has_target: bool,
            is_device_only: bool,
            device_type: u8,
            stream_index: i32,
        ) -> Result<UniquePtr<MlxArray>>;
        fn to_fp8(
            x: &MlxArray,
            has_target: bool,
            is_device_only: bool,
            device_type: u8,
            stream_index: i32,
        ) -> Result<UniquePtr<MlxArray>>;
    }
}
